test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs

```diff
@@ -2118,12 +2115,12 @@ public Task ResponseStatusCodeSetBeforeHttpContextDisposedRequestMalformedReadIg
                     // Don't wait to receive the response. Just close the socket.
                 }
 
-                connectionCloseWh.Set();
+                connectionCloseWh.TrySetResult(null);
 
                 // Write failed
-                await Assert.ThrowsAsync<TaskCanceledException>(async () => await writeTcs.Task);
+                await Assert.ThrowsAsync<TaskCanceledException>(async () => await writeTcs.Task).TimeoutAfter(TimeSpan.FromSeconds(15));
```

Comment:
Could be the switch to tcs. The continuation runs as soon as you call `SetResult` (i.e. in the same thread). We've had test deadlocks because of that before.